packages/pyright-internal/src/analyzer/binder.ts

```diff
@@ -255,7 +255,9 @@ export class Binder extends ParseTreeWalker {
         this._potentialPrivateSymbols.forEach((symbol, name) => {
             // If this symbol was found in the dunder all, don't mark it
             // as externally hidden.
-            if (!this._dunderAllNames?.some((sym) => sym === name)) {
+            const isInDunderAll = this._dunderAllNames?.some((sym) => sym === name);
+            const isTypedStubSymbol = this._fileInfo.isStubFile && symbol.getTypedDeclarations().length > 0;
+            if (!isInDunderAll && !isTypedStubSymbol) {
                 symbol.setIsExternallyHidden();
```

Comment:
I vaguely recall @heejaechang having to modify this code to ensure we could handle untyped libraries; I think this change should be safe given it explicitly applies to stubs only, but worth a second set of eyes.

Reply:
Yeah, I think this is safe for stubs.

Some of the typeshed stubs include symbols whose names would seem to indicate that they shouldn't be exported. For example, the `_exit` function in the `os.pyi` stub. That same stub also defines a symbol `_T`. I verified that mypy treats  `_exit` as an exported symbol but `_T` as a private (not exported) symbol. 

Note that this change results in a difference between a ".pyi" file and a pytyped ".py" file, at least by our current rules for pytype interpretation.

Reply:
Would this change only add the diagnostic (to warn uses of these things), and not affect the visibility of the symbols? That'd be my worry, that code that already uses something like `_exit` now getting changed to `Unknown`, when it's explicitly declared in the stub for a reason.

Reply:
There are two parts to the change in this PR. The first part avoids marking symbols as "externally hidden" if they come from a stub file and have a declared type. The second part adds a missing diagnostic check when accessing an "externally hidden" symbol from a module using a member access expression.

For clarity, here's how the behavior changes when accessing `os._exit` from the `os.pyi` stub.

Previous to this PR:
A1. We include `_exit` in the symbol table but mark it as "externally hidden".
A2. If you access this symbol using `from os import _exit`, you will see an error and the type will be marked unknown.
A3. If you access it using `import os` and `os._exit`, you will not see an error and the type will not be unknown.

Clearly the behavior between A2 and A3 is inconsistent and wrong.

With this PR:
B1. We include `_exit` in the symbol table and do not mark it as externally hidden because it comes from a stub file and it has a declared type (whereas `_T` does not have a declared type)
B2. If you import it using `from os import _exit`, you will not see an error
B3. If you import it using `import os` and `os._exit`, you will not see an error

So this change makes B2 and B3 consistent when the symbol comes from a stub.

I was previously concerned that this PR would change the visible behavior with py.typed libraries, but I've confirmed that it doesn't. That's because we don't enter private symbols 

Let's look at a different example where a symbol `foo` comes from a `py.typed` library, but `foo` is not publicly exported (e.g. it's imported from another module but not re-exported and not included in the `__all__` list).

Previous to this PR:
C1. We include `foo` in the symbol table and mark it externally hidden
C2. If you access this symbol using `from os import _exit`, you do not see an error and the type is known (this is a bug)
C3. If you access it using `import os` and `os._exit`, you do not see an error and the type is known (this is a bug)

With this PR:
D1. We include `foo` in the symbol table and mark it externally hidden
D2. If you access this symbol using `from os import _exit`, you do not see an error and the type is known (this is still a bug)
D3. If you access it using `import os` and `os._exit`, you see an error and the type is known

The visible behavior change is between C3 and D3, since users will now see an error. That's where we're likely to get some negative reaction to this change.

There's still a bug in case C2/D2. (I just discovered this as I was manually verifying these various cases.) I think we should fix this as part of this change. It's also likely to get some negative reactions, but I'd rather get all these fixes in place at the same time and "rip the bandaid off".